File: lib/src/models/part.dart
import 'package:meta/meta.dart';
import 'package:musicxml_parser/src/models/measure.dart';

/// Represents a single part (e.g., a single instrument or voice) within a musical score.
///
/// A part consists of a unique [id], an optional [name], and a list of [measures]
/// that make up the musical content of the part.
///
/// Instances are typically created via [PartBuilder].
/// Objects of this class are immutable.
@immutable
class Part {
  /// The unique identifier for this part within the score.
  final String id;

  /// The display name of the part (e.g., "Violin I", "Piano Left Hand").
  final String? name;

  /// The list of [Measure] objects that constitute this part.
  final List<Measure> measures;

  /// Creates a new [Part] instance.
  ///
  /// It is generally recommended to use [PartBuilder] for constructing [Part]
  /// objects, especially during parsing.
  const Part({
    required this.id,
    this.name,
    required this.measures,
  });

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Part &&
          runtimeType == other.runtimeType &&
          id == other.id &&
          name == other.name &&
          // Note: DeepCollectionEquality might be needed for measures if deep comparison is critical
          // For now, relying on list's default equality (identity or element-wise if elements override ==)
          measures == other.measures;

  @override
  int get hashCode => id.hashCode ^ (name?.hashCode ?? 0) ^ measures.hashCode;

  @override
  String toString() =>
      'Part{id: $id, name: $name, measures: ${measures.length}}';
}

/// Builder for creating [Part] objects incrementally.
///
/// This builder is useful during the parsing process for MusicXML `<part>` elements,
/// allowing measures to be added one by one as they are parsed.
/// The [build] method finalizes the part construction.
///
/// Example:
/// ```dart
/// final partBuilder = PartBuilder("P1", line: 20)..setName("Flute");
/// partBuilder.addMeasure(firstMeasure);
/// partBuilder.addMeasure(secondMeasure);
/// final Part flutePart = partBuilder.build();
/// ```
class PartBuilder {
  /// The ID of the part being built.
  final String _id;
  String? _name;
  List<Measure> _measures = [];

  /// Line number in the XML for error reporting context.
  final int? _line;

  /// Additional context for error reporting.
  final Map<String, dynamic>? _context;

  /// Creates a [PartBuilder] for a part with the given [id].
  ///
  /// [line] and [context] can be provided for more detailed error
  /// messages if validation (not currently implemented in builder) were to fail.
  PartBuilder(this._id, {int? line, Map<String, dynamic>? context})
      : _line = line,
        _context = context;

  /// Sets the name of the part.
  PartBuilder setName(String? name) {
    _name = name;
    return this;
  }

  /// Sets all measures for the part.
  PartBuilder setMeasures(List<Measure> measures) {
    _measures = measures;
    return this;
  }

  /// Adds a single [Measure] to the part.
  PartBuilder addMeasure(Measure measure) {
    _measures.add(measure);
    return this;
  }

  /// Builds the [Part] instance.
  ///
  /// This method constructs the [Part] object from the properties set
  /// on the builder. Currently, it performs minimal validation itself,
  /// relying on the parser to ensure required fields like ID are present.
  Part build() {
    if (_id.isEmpty) {
      // This should ideally be caught by the PartParser before even creating the builder
      // but serves as a safeguard or location for future builder-specific validation.
      throw ArgumentError('Part ID cannot be empty.');
    }
    return Part(
      id: _id,
      name: _name,
      measures: _measures,
    );
  }
}

File: csharp/MusicXMLParser/Models/Part.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace MusicXMLParser.Models
{
    /// <summary>
    /// Represents a single part (e.g., a single instrument or voice) within a musical score.
    /// </summary>
    /// <remarks>
    /// A part consists of a unique <see cref="Id"/>, an optional <see cref="Name"/>, and a list of <see cref="Measures"/>
    /// that make up the musical content of the part.
    /// Instances are typically created via <see cref="PartBuilder"/>.
    /// Objects of this class are immutable once created by the builder.
    /// </remarks>
    public class Part
    {
        /// <summary>
        /// The unique identifier for this part within the score.
        /// </summary>
        public string Id { get; }

        /// <summary>
        /// The display name of the part (e.g., "Violin I", "Piano Left Hand").
        /// </summary>
        public string Name { get; }

        /// <summary>
        /// The list of <see cref="Measure"/> objects that constitute this part.
        /// </summary>
        public List<Measure> Measures { get; }

        /// <summary>
        /// Creates a new <see cref="Part"/> instance.
        /// </summary>
        /// <remarks>
        /// It is generally recommended to use <see cref="PartBuilder"/> for constructing <see cref="Part"/>
        /// objects, especially during parsing.
        /// </remarks>
        public Part(string id, string name, List<Measure> measures)
        {
            Id = id;
            Name = name;
            Measures = measures;
        }

        public override bool Equals(object obj)
        {
            if (obj is Part other)
            {
                return Id == other.Id &&
                       Name == other.Name &&
                       Measures.SequenceEqual(other.Measures);
            }
            return false;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Id, Name, Measures);
        }

        public override string ToString() =>
            $"Part{{Id: {Id}, Name: {Name}, Measures: {Measures.Count}}}";
    }

    /// <summary>
    /// Builder for creating <see cref="Part"/> objects incrementally.
    /// </summary>
    /// <remarks>
    /// This builder is useful during the parsing process for MusicXML &lt;part&gt; elements,
    /// allowing measures to be added one by one as they are parsed.
    /// The <see cref="Build"/> method finalizes the part construction.
    /// </remarks>
    /// <example>
    /// <code>
    /// var partBuilder = new PartBuilder("P1").SetName("Flute");
    /// partBuilder.AddMeasure(firstMeasure);
    /// partBuilder.AddMeasure(secondMeasure);
    /// Part flutePart = partBuilder.Build();
    /// </code>
    /// </example>
    public class PartBuilder
    {
        private readonly string _id;
        private string _name;
        private List<Measure> _measures = new List<Measure>();

        /// <summary>
        /// Creates a <see cref="PartBuilder"/> for a part with the given <paramref name="id"/>.
        /// </summary>
        public PartBuilder(string id)
        {
            if (string.IsNullOrEmpty(id))
            {
                throw new ArgumentException("Part ID cannot be null or empty.", nameof(id));
            }
            _id = id;
        }

        /// <summary>
        /// Sets the name of the part.
        /// </summary>
        public PartBuilder SetName(string name)
        {
            _name = name;
            return this;
        }

        /// <summary>
        /// Sets all measures for the part.
        /// </summary>
        public PartBuilder SetMeasures(List<Measure> measures)
        {
            _measures = measures ?? new List<Measure>();
            return this;
        }

        /// <summary>
        /// Adds a single <see cref="Measure"/> to the part.
        /// </summary>
        public PartBuilder AddMeasure(Measure measure)
        {
            if (measure != null)
            {
                _measures.Add(measure);
            }
            return this;
        }

        /// <summary>
        /// Builds the <see cref="Part"/> instance.
        /// </summary>
        /// <remarks>
        /// This method constructs the <see cref="Part"/> object from the properties set
        /// on the builder.
        /// </remarks>
        public Part Build()
        {
            return new Part(_id, _name, _measures);
        }
    }
}
